core/azure-core/src/main/java/com/azure/core/util/Context.java

```diff
@@ -77,4 +95,62 @@ public Context addData(Object key, Object value) {
         }
         return Optional.empty();
     }
+
+    @Override
+    public void close() {
+        // Because of how we use context, we can not simply call threadLocal.remove() here, as
+        // the context created at the start of a try-with-resources will not necessarily be the one
+        // that needs to have close() called on it, and from that context we have no means on finding
+        // the right context, except for diving into the thread reflectively.
+        Context ctx = getThreadLocalContext();
+        if (ctx != null && ctx.threadLocal != null) {
+            ctx.threadLocal.remove();
+        }
+    }
+
+    // FIXME temporary toString for debugging - delete before merging
+    @Override
+    public String toString() {
+        if (key == null && value == null) {
+            return "[Context NONE, parent=" + parent + "]";
+        }
+        return "[Context key='" + key + "', value='" + value + "', parent=" + parent + "]";
+    }
+
+    public static Context getThreadLocalContext() {
+        Context result = Context.NONE;
+
+        try {
+            if (!reflectionInitiated) {
+                threadLocalsField = Thread.class.getDeclaredField("threadLocals");
```

Comment:
As mentioned there are security manager implications with reflections, would the be a fallback if this doesn't work?

In the recent storage API overview we removed context as a parameter in the T client methods but continue to include them in the Response<T> clients. Does that feel like a better compromise?